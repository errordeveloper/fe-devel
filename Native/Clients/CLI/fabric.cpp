/*
 *  Copyright 2010-2011 Fabric Technologies Inc. All rights reserved.
 */

#include <Fabric/Clients/CLI/ClientFactory.h>
#include <Fabric/Core/Util/Log.h>
#include <Fabric/Core/Build.h>

#include <time.h>

namespace Fabric
{
  namespace CLI
  {
    extern "C" FABRIC_CLI_EXPORT void init( v8::Handle<v8::Object> target )
    {
      FABRIC_LOG( "%s version %s", Fabric::buildName, Fabric::buildFullVersion );
      struct tm const *lt = localtime( &Fabric::buildExpiry );
      char buf[1024];
      strftime( buf, 1024, "This build of Fabric will expire on %Y-%m-%d at %H:%M:%S", lt );
      FABRIC_LOG( "%s", buf );

      v8::HandleScope handleScope;
      target->Set( v8::String::New("createClient"), CreateClientV8Function() );
    }
  };
};
