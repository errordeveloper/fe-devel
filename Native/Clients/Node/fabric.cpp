/*
 *  Copyright 2010-2012 Fabric Engine Inc. All rights reserved.
 */

#include <Fabric/Clients/Node/ClientFactory.h>
#include <Fabric/Base/Util/Log.h>
#include <Fabric/Core/Build.h>

#include <llvm/Support/Threading.h>
#include <time.h>
#include <stdlib.h>

namespace Fabric
{
  namespace CLI
  {
    extern "C" FABRIC_CLI_EXPORT void init( v8::Handle<v8::Object> target )
    {
      FABRIC_LOG( "%s version %s", Fabric::buildName, Fabric::buildFullVersion );

      llvm::llvm_start_multithreaded();
      if ( !llvm::llvm_is_multithreaded() )
      {
        FABRIC_LOG( "LLVM not compiled with multithreading enabled; aborting" );
        abort();
      }

      v8::HandleScope handleScope;
      target->Set( v8::String::New("createClient"), CreateClientV8Function() );
    }
  }
}
