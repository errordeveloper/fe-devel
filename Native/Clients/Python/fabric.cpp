/*
 *  Copyright 2010-2012 Fabric Engine Inc. All rights reserved.
 */

#include <Fabric/Base/Util/SimpleString.h>
#include <Fabric/Base/Exception.h>
#include <Fabric/Clients/Python/ClientWrap.h>
#include <Fabric/Base/Util/Log.h>
#include <Fabric/Core/Build.h>

#include <map>
#include <time.h>
#include <llvm/Support/Threading.h>

//#define FABRIC_PYTHON_DEBUG

namespace Fabric
{
  namespace Python
  {
    extern "C" FABRIC_CLI_EXPORT void initialize()
    {
      FABRIC_LOG( "%s version %s", Fabric::buildName, Fabric::buildFullVersion );

      llvm::llvm_start_multithreaded();
      if ( !llvm::llvm_is_multithreaded() )
      {
        FABRIC_LOG( "LLVM not compiled with multithreading enabled; aborting" );
        abort();
      }
    }

    extern "C" FABRIC_CLI_EXPORT void createClient( void **ptr, char const *opts )
    {
      ClientWrap *client = new ClientWrap( opts );

#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "calling createClient: %lx", client);
#endif

      client->retain();
      *ptr = client;
    }
    
    extern "C" FABRIC_CLI_EXPORT void jsonExec( void *client_, char const *data, size_t length, const char **result )
    {
#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "calling jsonExec: %lx", client_ );
#endif

      ClientWrap *client( static_cast<ClientWrap *>( client_ ) );

#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "sending to jsonExec: %s", data );
#endif

      client->jsonExecAndAllocCStr( data, length, result );

#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "received from jsonExec: %s", *result );
#endif
    }

    extern "C" FABRIC_CLI_EXPORT void freeString( void *client_, void *string )
    {
#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "calling freeString: %lx (%lx)", client_, string );
#endif

      ClientWrap *client( static_cast<ClientWrap *>( client_ ) );
      client->freeJsonCStr( (const char *)string );
    }

    extern "C" FABRIC_CLI_EXPORT void setJSONNotifyCallback(
        void *client_,
        void (*callback)(const char *)
        )
    {
#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "calling setJSONNotifyCallback: %lx", client_ );
#endif

      ClientWrap *client( static_cast<ClientWrap *>( client_ ) );
      client->setJSONNotifyCallback( callback );
    }

    extern "C" FABRIC_CLI_EXPORT void freeClient( void *client_ )
    {
#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "calling close: %lx", client_ );
#endif

      ClientWrap *client( static_cast<ClientWrap *>( client_ ) );
      client->release();
    }

    extern "C" FABRIC_CLI_EXPORT void runScheduledCallbacks( void *client_ )
    {
#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "calling runScheduledCallbacks: %lx", client_ );
#endif

      ClientWrap *client( static_cast<ClientWrap *>( client_ ) );
      client->runScheduledCallbacks();
    }
  };
};
