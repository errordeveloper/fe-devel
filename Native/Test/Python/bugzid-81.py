#
#  Copyright 2010-2012 Fabric Engine Inc. All rights reserved.
#

import fabric
F = fabric.createClient()

# create a new dgnode
dgnode1 = F.DG.createNode("myNode1")
dgnode2 = F.DG.createNode("myNode2")

# add some members
dgnode1.addMember('a','Scalar')
dgnode1.addMember('b','Scalar')
dgnode1.addMember('result','Scalar')

dgnode2.addMember('result','Scalar')

# increase the slicecount of the nodes
count = 1000000
dgnode1.resize(count)
dgnode2.resize(count)

# create dependencies between the nodes
dgnode2.setDependency(dgnode1, 'otherNode')

# create three operators, one to initiate the data, one to add, and one to mul
operatorInit = F.DG.createOperator("initiate")
operatorInit.setSourceCode(
  'operator initiate(in Size index, io Scalar a, io Scalar b) {\n'+
  '  a = 12.5 * Scalar(index);\n'+
  '  b = 1.73 * Scalar(index);\n'+
  '}\n')
operatorInit.setEntryPoint('initiate')
if len( operatorInit.getErrors() ) > 0:
  if len( operatorInit.getDiagnostics() ) > 0:
    print(fabric.stringify(opreatorInit.getDiagnostics()))

operatorAdd = F.DG.createOperator("add")
operatorAdd.setSourceCode(
  'operator add(io Scalar a, io Scalar b, io Scalar result) {\n'+
  '  result = a + b;\n'+
  '}\n')
operatorAdd.setEntryPoint('add')

if len( operatorAdd.getErrors() ) > 0:
  if len( operatorAdd.getDiagnostics() ) > 0:
    print(fabric.stringify(opreatorAdd.getDiagnostics()))

operatorMul = F.DG.createOperator("operatorMul")
operatorMul.setSourceCode(
  'operator mul(io Scalar input<>, io Scalar result, in Size index) {\n'+
  '  result = 3.0 * input[index];\n'+
  '}\n')
operatorMul.setEntryPoint('mul')

if len( operatorMul.getErrors() ) > 0:
  if len( operatorMul.getDiagnostics() ) > 0:
    print(fabric.stringify(opreatorMul.getDiagnostics()))

# create bindings between the nodes and the operator
binding1init = F.DG.createBinding()
binding1init.setOperator(operatorInit)
binding1init.setParameterLayout([
  'self.index',
  'self.a',
  'self.b'
])
binding1add = F.DG.createBinding()
binding1add.setOperator(operatorAdd)
binding1add.setParameterLayout([
  'self.a',
  'self.b',
  'self.result'
])
binding2 = F.DG.createBinding()
binding2.setOperator(operatorMul)
binding2.setParameterLayout([
  'otherNode.result<>',
  'self.result',
  'self.index'
])

# append the new binding to the node
dgnode1.bindings.append(binding1init)
dgnode1.bindings.append(binding1add)
if len( dgnode1.getErrors() ) > 0:
  print(fabric.stringify(dgnode1.getErrors()))
dgnode2.bindings.append(binding2)
if len( dgnode2.getErrors() ) > 0:
  print(fabric.stringify(dgnode2.getErrors()))

# evaluate the node!
dgnode2.evaluate()

# query the node's data (only for certain indices)
displayCount = 100 if count > 100 else count
indices = []
for i in range( 0, displayCount ):
  indices.append(i)

print(fabric.stringify(dgnode1.getSlicesBulkData(indices)))
print(fabric.stringify(dgnode2.getSlicesBulkData(indices)))

F.close()
