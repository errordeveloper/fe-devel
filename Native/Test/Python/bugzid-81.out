[[{"a": 0}, {"b": 0}, {"result": 0}], [{"a": "12.500"}, {"b": "1.730"}, {"result": "14.230"}], [{"a": 25}, {"b": "3.460"}, {"result": "28.460"}], [{"a": "37.500"}, {"b": "5.190"}, {"result": "42.690"}], [{"a": 50}, {"b": "6.920"}, {"result": "56.920"}], [{"a": "62.500"}, {"b": "8.650"}, {"result": "71.150"}], [{"a": 75}, {"b": "10.380"}, {"result": "85.380"}], [{"a": "87.500"}, {"b": "12.110"}, {"result": "99.610"}], [{"a": 100}, {"b": "13.840"}, {"result": "113.840"}], [{"a": "112.500"}, {"b": "15.570"}, {"result": "128.070"}], [{"a": 125}, {"b": "17.300"}, {"result": "142.300"}], [{"a": "137.500"}, {"b": "19.030"}, {"result": "156.530"}], [{"a": 150}, {"b": "20.760"}, {"result": "170.760"}], [{"a": "162.500"}, {"b": "22.490"}, {"result": "184.990"}], [{"a": 175}, {"b": "24.220"}, {"result": "199.220"}], [{"a": "187.500"}, {"b": "25.950"}, {"result": "213.450"}], [{"a": 200}, {"b": "27.680"}, {"result": "227.680"}], [{"a": "212.500"}, {"b": "29.410"}, {"result": "241.910"}], [{"a": 225}, {"b": "31.140"}, {"result": "256.140"}], [{"a": "237.500"}, {"b": "32.870"}, {"result": "270.370"}], [{"a": 250}, {"b": "34.600"}, {"result": "284.600"}], [{"a": "262.500"}, {"b": "36.330"}, {"result": "298.830"}], [{"a": 275}, {"b": "38.060"}, {"result": "313.060"}], [{"a": "287.500"}, {"b": "39.790"}, {"result": "327.290"}], [{"a": 300}, {"b": "41.520"}, {"result": "341.520"}], [{"a": "312.500"}, {"b": "43.250"}, {"result": "355.750"}], [{"a": 325}, {"b": "44.980"}, {"result": "369.980"}], [{"a": "337.500"}, {"b": "46.710"}, {"result": "384.210"}], [{"a": 350}, {"b": "48.440"}, {"result": "398.440"}], [{"a": "362.500"}, {"b": "50.170"}, {"result": "412.670"}], [{"a": 375}, {"b": "51.900"}, {"result": "426.900"}], [{"a": "387.500"}, {"b": "53.630"}, {"result": "441.130"}], [{"a": 400}, {"b": "55.360"}, {"result": "455.360"}], [{"a": "412.500"}, {"b": "57.090"}, {"result": "469.590"}], [{"a": 425}, {"b": "58.820"}, {"result": "483.820"}], [{"a": "437.500"}, {"b": "60.550"}, {"result": "498.050"}], [{"a": 450}, {"b": "62.280"}, {"result": "512.280"}], [{"a": "462.500"}, {"b": "64.010"}, {"result": "526.510"}], [{"a": 475}, {"b": "65.740"}, {"result": "540.740"}], [{"a": "487.500"}, {"b": "67.470"}, {"result": "554.970"}], [{"a": 500}, {"b": "69.200"}, {"result": "569.200"}], [{"a": "512.500"}, {"b": "70.930"}, {"result": "583.430"}], [{"a": 525}, {"b": "72.660"}, {"result": "597.660"}], [{"a": "537.500"}, {"b": "74.390"}, {"result": "611.890"}], [{"a": 550}, {"b": "76.120"}, {"result": "626.120"}], [{"a": "562.500"}, {"b": "77.850"}, {"result": "640.350"}], [{"a": 575}, {"b": "79.580"}, {"result": "654.580"}], [{"a": "587.500"}, {"b": "81.310"}, {"result": "668.810"}], [{"a": 600}, {"b": "83.040"}, {"result": "683.040"}], [{"a": "612.500"}, {"b": "84.770"}, {"result": "697.270"}], [{"a": 625}, {"b": "86.500"}, {"result": "711.500"}], [{"a": "637.500"}, {"b": "88.230"}, {"result": "725.730"}], [{"a": 650}, {"b": "89.960"}, {"result": "739.960"}], [{"a": "662.500"}, {"b": "91.690"}, {"result": "754.190"}], [{"a": 675}, {"b": "93.420"}, {"result": "768.420"}], [{"a": "687.500"}, {"b": "95.150"}, {"result": "782.650"}], [{"a": 700}, {"b": "96.880"}, {"result": "796.880"}], [{"a": "712.500"}, {"b": "98.610"}, {"result": "811.110"}], [{"a": 725}, {"b": "100.340"}, {"result": "825.340"}], [{"a": "737.500"}, {"b": "102.070"}, {"result": "839.570"}], [{"a": 750}, {"b": "103.800"}, {"result": "853.800"}], [{"a": "762.500"}, {"b": "105.530"}, {"result": "868.030"}], [{"a": 775}, {"b": "107.260"}, {"result": "882.260"}], [{"a": "787.500"}, {"b": "108.990"}, {"result": "896.490"}], [{"a": 800}, {"b": "110.720"}, {"result": "910.720"}], [{"a": "812.500"}, {"b": "112.450"}, {"result": "924.950"}], [{"a": 825}, {"b": "114.180"}, {"result": "939.180"}], [{"a": "837.500"}, {"b": "115.910"}, {"result": "953.410"}], [{"a": 850}, {"b": "117.640"}, {"result": "967.640"}], [{"a": "862.500"}, {"b": "119.370"}, {"result": "981.870"}], [{"a": 875}, {"b": "121.100"}, {"result": "996.100"}], [{"a": "887.500"}, {"b": "122.830"}, {"result": "1010.330"}], [{"a": 900}, {"b": "124.560"}, {"result": "1024.560"}], [{"a": "912.500"}, {"b": "126.290"}, {"result": "1038.790"}], [{"a": 925}, {"b": "128.020"}, {"result": "1053.020"}], [{"a": "937.500"}, {"b": "129.750"}, {"result": "1067.250"}], [{"a": 950}, {"b": "131.480"}, {"result": "1081.480"}], [{"a": "962.500"}, {"b": "133.210"}, {"result": "1095.710"}], [{"a": 975}, {"b": "134.940"}, {"result": "1109.940"}], [{"a": "987.500"}, {"b": "136.670"}, {"result": "1124.170"}], [{"a": 1000}, {"b": "138.400"}, {"result": "1138.400"}], [{"a": "1012.500"}, {"b": "140.130"}, {"result": "1152.630"}], [{"a": 1025}, {"b": "141.860"}, {"result": "1166.860"}], [{"a": "1037.500"}, {"b": "143.590"}, {"result": "1181.090"}], [{"a": 1050}, {"b": "145.320"}, {"result": "1195.320"}], [{"a": "1062.500"}, {"b": "147.050"}, {"result": "1209.550"}], [{"a": 1075}, {"b": "148.780"}, {"result": "1223.780"}], [{"a": "1087.500"}, {"b": "150.510"}, {"result": "1238.010"}], [{"a": 1100}, {"b": "152.240"}, {"result": "1252.240"}], [{"a": "1112.500"}, {"b": "153.970"}, {"result": "1266.470"}], [{"a": 1125}, {"b": "155.700"}, {"result": "1280.700"}], [{"a": "1137.500"}, {"b": "157.430"}, {"result": "1294.930"}], [{"a": 1150}, {"b": "159.160"}, {"result": "1309.160"}], [{"a": "1162.500"}, {"b": "160.890"}, {"result": "1323.390"}], [{"a": 1175}, {"b": "162.620"}, {"result": "1337.620"}], [{"a": "1187.500"}, {"b": "164.350"}, {"result": "1351.850"}], [{"a": 1200}, {"b": "166.080"}, {"result": "1366.080"}], [{"a": "1212.500"}, {"b": "167.810"}, {"result": "1380.310"}], [{"a": 1225}, {"b": "169.540"}, {"result": "1394.540"}], [{"a": "1237.500"}, {"b": "171.270"}, {"result": "1408.770"}]]
[[{"result": 0}], [{"result": "42.690"}], [{"result": "85.380"}], [{"result": "128.070"}], [{"result": "170.760"}], [{"result": "213.450"}], [{"result": "256.140"}], [{"result": "298.830"}], [{"result": "341.520"}], [{"result": "384.210"}], [{"result": "426.900"}], [{"result": "469.590"}], [{"result": "512.280"}], [{"result": "554.970"}], [{"result": "597.660"}], [{"result": "640.350"}], [{"result": "683.040"}], [{"result": "725.730"}], [{"result": "768.420"}], [{"result": "811.110"}], [{"result": "853.800"}], [{"result": "896.490"}], [{"result": "939.180"}], [{"result": "981.870"}], [{"result": "1024.560"}], [{"result": "1067.250"}], [{"result": "1109.940"}], [{"result": "1152.630"}], [{"result": "1195.320"}], [{"result": "1238.010"}], [{"result": "1280.700"}], [{"result": "1323.390"}], [{"result": "1366.080"}], [{"result": "1408.770"}], [{"result": "1451.460"}], [{"result": "1494.150"}], [{"result": "1536.840"}], [{"result": "1579.530"}], [{"result": "1622.220"}], [{"result": "1664.910"}], [{"result": "1707.600"}], [{"result": "1750.290"}], [{"result": "1792.980"}], [{"result": "1835.670"}], [{"result": "1878.360"}], [{"result": "1921.050"}], [{"result": "1963.740"}], [{"result": "2006.430"}], [{"result": "2049.120"}], [{"result": "2091.810"}], [{"result": "2134.500"}], [{"result": "2177.190"}], [{"result": "2219.880"}], [{"result": "2262.570"}], [{"result": "2305.260"}], [{"result": "2347.950"}], [{"result": "2390.640"}], [{"result": "2433.330"}], [{"result": "2476.020"}], [{"result": "2518.710"}], [{"result": "2561.400"}], [{"result": "2604.090"}], [{"result": "2646.780"}], [{"result": "2689.470"}], [{"result": "2732.160"}], [{"result": "2774.850"}], [{"result": "2817.540"}], [{"result": "2860.230"}], [{"result": "2902.920"}], [{"result": "2945.610"}], [{"result": "2988.300"}], [{"result": "3030.990"}], [{"result": "3073.680"}], [{"result": "3116.370"}], [{"result": "3159.060"}], [{"result": "3201.750"}], [{"result": "3244.440"}], [{"result": "3287.130"}], [{"result": "3329.820"}], [{"result": "3372.510"}], [{"result": "3415.200"}], [{"result": "3457.890"}], [{"result": "3500.580"}], [{"result": "3543.270"}], [{"result": "3585.960"}], [{"result": "3628.650"}], [{"result": "3671.340"}], [{"result": "3714.030"}], [{"result": "3756.720"}], [{"result": "3799.410"}], [{"result": "3842.100"}], [{"result": "3884.790"}], [{"result": "3927.480"}], [{"result": "3970.170"}], [{"result": "4012.860"}], [{"result": "4055.550"}], [{"result": "4098.240"}], [{"result": "4140.930"}], [{"result": "4183.620"}], [{"result": "4226.310"}]]
